src/libmongoc/src/mongoc/mongoc-cursor-find.c

```diff
@@ -45,12 +45,17 @@ _prime (mongoc_cursor_t *cursor)
    mongoc_server_stream_cleanup (server_stream);
 
    /* set all mongoc_impl_t function pointers. */
-   if (
-      /* Server version 5.1 and newer do not support OP_QUERY. */
-      wire_version > WIRE_VERSION_5_0 ||
-      /* Fallback to legacy OP_QUERY wire protocol messages if exhaust cursor
-         requested. */
-      !_mongoc_cursor_get_opt_bool (cursor, MONGOC_CURSOR_EXHAUST)) {
+   const bool sharded = _mongoc_topology_get_type (cursor->client->topology) ==
+                        MONGOC_TOPOLOGY_SHARDED;
+   const bool is_exhaust =
+      _mongoc_cursor_get_opt_bool (cursor, MONGOC_CURSOR_EXHAUST);
+   const bool server_supports_getmore_no_exhaust =
+      wire_version >= WIRE_VERSION_3_6;
+   const bool server_supports_getmore_exhaust =
+      sharded ? wire_version >= WIRE_VERSION_MONGOS_EXHAUST
+              : wire_version >= WIRE_VERSION_4_2;
+   if ((is_exhaust && server_supports_getmore_exhaust) ||
+       (!is_exhaust && server_supports_getmore_no_exhaust)) {
```

Comment:
Same checks as in `_getmore_type ()` in mongoc-cursor-cmd.c. Consider refactoring both into a common helper function.